website/blog/2025/01-29-superscraper/index.md

```diff
@@ -0,0 +1,189 @@
+
+```js
+const crawlers = new Map<string, PlaywrightCrawler>();
+```
+
+Here’s a part of the code that gets executed when a new request from the user arrives; if the crawler for this proxy configuration exists in the map, it will be used. Otherwise, a new crawler gets created. Then, we add the request to the crawler’s queue so it can be processed.
+
+```js
+const key = JSON.stringify(crawlerOptions); 
+const crawler = crawlers.has(key) ? crawlers.get(key)! : await createAndStartCrawler(crawlerOptions);
+
+await crawler.requestQueue!.addRequest(request);
+```
+
+The function below initializes new crawlers with predefined settings and behaviors. Each crawler utilizes its own in-memory queue created with the `MemoryStorage` client. This approach is used for two key reasons:
+
+1. **Performance**: In-memory queues are faster, and there's no need to persist them when SuperScraper migrates.
+2. **Isolation**: Using a separate queue prevents interference with the shared default queue of the SuperScraper Actor, avoiding potential bugs when multiple crawlers use it simultaneously.
+
+```js
+export const createAndStartCrawler = async (crawlerOptions: CrawlerOptions = DEFAULT_CRAWLER_OPTIONS) => {
+    const client = new MemoryStorage();
+    const queue = await RequestQueue.open(undefined, { storageClient: client });
+
+    const proxyConfig = await Actor.createProxyConfiguration(crawlerOptions.proxyConfigurationOptions);
+
+    const crawler = new PlaywrightCrawler({
+        keepAlive: true,
+        proxyConfiguration: proxyConfig,
+        maxRequestRetries: 4,
+        requestQueue: queue,
+    });
+};
+```
+
+At the end of the function, we start the crawler and log a message if it terminates for any reason. Next, we add the newly created crawler to the key-value map containing all crawlers, and finally, we return the crawler.
+
+```js
+crawler.run().then(
+    () => log.warning(`Crawler ended`, crawlerOptions),
+    () => { }
+);
+
+crawlers.set(JSON.stringify(crawlerOptions), crawler);
+
+log.info('Crawler ready 🚀', crawlerOptions);
+
+return crawler;
+```
+
+### Mapping standby HTTP requests to Crawlee requests
+
+When creating the server, it accepts a request listener function that takes two arguments: the user’s request and a response object. The response object is used to send scraped data back to the user. These response objects are stored in a key-value map to so they can be accessed later in the code. The key is a randomly generated string shared between the request and its corresponding response object, it is used as `request.uniqueKey`.
+
+```js
+const responses = new Map<string, ServerResponse>();
+```
+
+**Saving response objects**
+
+The following function stores a response object in the key-value map:
+
+```js
```

Comment:
the code is TS since you have type hints in there, not JS, I can see the highlighting wouldn't be happy about this. most (if not all) examples here should use `ts`.